### DISCUSSION TITLE
What is the recommended collaborative project workflow (especially for iOS)?

### DISCUSSION THREAD

**Discussion by katwdev91** [axmolengine/axmol#1887] [Q&A] [answered]
We are a small team and would like to use the engine for our project (migrating from Cocos2dx). We've been playing with the engine for a while and tried to setup a git repo for our project. However we are having issues with setting up the proejct and still don't understand the expected workflow for Axmol. We're using Git + Axmol 2.1.2 + Cocoapods.
 
Our specific questions are:
- Is there a recommended .gitignore to use for Axmol project? We've been creating our own but would love to know if there's a recommended one.
- For Android it's very straightforward as it's simply opening the proj.android with Android Studio (same as original Cocos2dx). However for iOS, should we run `axmol build` and generate a new Xcode project every time we want to build for iOS? 
- Should the generated iOS build folder be tracked by Git? How would one set this up with Git for multiple developers on the same project? 
- We're thinking about avoiding generating / overriding the build folder repeatedly since we're also using Cocoapods for iOS.

Any direction pointing is appreciated. Thanks!

**Comment by rh101** [comment, marked as answer]
> However we are having issues with setting up the proejct and still don't understand the expected workflow for Axmol. We're using Git + Axmol 2.1.2 + Cocoapods.

As a personal opinion, you should use the latest dev branch of Axmol (which will eventually be released as 2.1.3), since the changes are significant.  

> Our specific questions are:
> 
>     * Is there a recommended .gitignore to use for Axmol project? We've been creating our own but would love to know if there's a recommended one.

A `.gitignore` that handles Visual Studio, Android and Xcode would work, along with a few extra additions like these:
```
# ignore all build
/build
/build_*

# ignore ninja build dir
/out
/out*

# external libs zip
**/simulator/

# New .cxx temp folder
**/.cxx

**/.idea

**/SharedLoader.java

.DS_Store
._*
._*.*
xcuserdata/
project.xcworkspace

.vscode/
/run.bat
/run.bat.txt
```

>     * For Android it's very straightforward as it's simply opening the proj.android with Android Studio (same as original Cocos2dx). However for iOS, should we run `axmol build` and generate a new Xcode project every time we want to build for iOS?

Android Studio automatically runs cmake in the background, which is why you don't need to run it manually via the `axmol build` command.  For the other platforms, you need to generate the projects yourself, so you either run `axmol build ...` or equivalent `cmake ...`
command for this task.

You only need to generate the Xcode project once, and only re-run the command again if you modify anything, such as new resources, new source files, or any other edits to `CMakeLists.txt`.  This will regenerate the project with the correct references to the new items.  Note that any edits you do to the project inside Xcode will be lost when you regenerate the project via cmake, so any changes you need to make to the project must be done in the `CMakeLists.txt`.  This is not specifically related to Axmol, but rather just how cmake builds work.

>     * Should the generated iOS build folder be tracked by Git? How would one set this up with Git for multiple developers on the same project?

No, you should not track the build folders in git. Any developer that clones a repository should simply generate the build files in their own development environment via cmake or `axmol build`.   

>     * We're thinking about avoiding generating / overriding the build folder repeatedly since we're also using Cocoapods for iOS.

The use of Cocoapods makes no difference, since it is compatible with the cmake workflow.  There are other posts on this discussion board regarding how to set this up correctly.  For instance, this thread: https://github.com/axmolengine/axmol/discussions/1817

**Reply by katwdev91** [reply]
Thank you so much for the answer! It makes more sense now. We found out the version 2.1.3 (beta2) today and gave it a try. The new `-i` option is really helpful since we might need to change a few things in the engine for the project, which is exactly what we needed. Really excited to see the new release. Will definite use the 2.1.3 stable version when it's out. Thanks again for the answers!
